tee_logger.go

```diff
@@ -0,0 +1,120 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+// TeeLogger creates a Logger that duplicates its log calls to two or
+// more loggers. It is similar to io.MultiWriter.
+//
+// For each logging level methad (.Debug, .Info, etc), the TeeLogger calls each
+// sub-logger's level method.
+//
+// Exceptions are made for the Fatal and Panic methods: the returned logger
+// calls .Log(FatalLevel, ...) and .Log(PanicLevel, ...). Only after all
+// sub-loggers have received the message, then the TeeLogger terminates the
+// process (using os.Exit or panic() per usual semantics).
+//
+// DFatal is handled similarly to Fatal and Panic, since it is not actually a
+// level; each sub-logger's DFatal method dynamically chooses to either cal
+// Error or Fatal.
+//
+// Check returns a CheckedMessage chain of any OK CheckedMessages returned by
+// all sub-loggers. The returned message is OK if any of the sub-messages are.
+// An exception is made for FatalLevel and PanicLevel, where a CheckedMessage
+// is returned against the TeeLogger itself. This is so that
+// tlog.Check(PanicLevel, ...).Write(...) is equivalent to tlog.Panic(...)
+// (likewise for FatalLevel).
+func TeeLogger(logs ...Logger) Logger {
+	switch len(logs) {
+	case 0:
+		return nil
+	case 1:
+		return logs[0]
+	default:
+		return multiLogger(logs)
+	}
+}
+
+type multiLogger []Logger
+
+func (ml multiLogger) Log(lvl Level, msg string, fields ...Field) {
+	ml.log(lvl, msg, fields)
+}
+
+func (ml multiLogger) Debug(msg string, fields ...Field) {
+	ml.log(DebugLevel, msg, fields)
+}
+
+func (ml multiLogger) Info(msg string, fields ...Field) {
+	ml.log(InfoLevel, msg, fields)
+}
+
+func (ml multiLogger) Warn(msg string, fields ...Field) {
+	ml.log(WarnLevel, msg, fields)
+}
+
+func (ml multiLogger) Error(msg string, fields ...Field) {
+	ml.log(ErrorLevel, msg, fields)
+}
+
+func (ml multiLogger) Panic(msg string, fields ...Field) {
+	ml.log(PanicLevel, msg, fields)
+	panic(msg)
+}
+
+func (ml multiLogger) Fatal(msg string, fields ...Field) {
+	ml.log(FatalLevel, msg, fields)
+	_exit(1)
+}
+
+func (ml multiLogger) log(lvl Level, msg string, fields []Field) {
+	for _, log := range ml {
+		log.Log(lvl, msg, fields...)
+	}
+}
+
+func (ml multiLogger) DFatal(msg string, fields ...Field) {
+	for _, log := range ml {
+		log.DFatal(msg, fields...)
+	}
+}
+
+func (ml multiLogger) With(fields ...Field) Logger {
+	ml = append([]Logger(nil), ml...)
+	for i, log := range ml {
+		ml[i] = log.With(fields...)
+	}
+	return multiLogger(ml)
```

Comment:
Since `multiLogger` exposes the higher-perf `Logger` API, let's golf away the copy here. Can we use a `for i :=` loop and `ml[i] = ml[i].With(fields...)` inside the loop?

(I'll open a separate PR to do the same throughout the core logger.)